### ISSUE TITLE
[BUG]: ERROR TypeError: Cannot read property 'execute' of undefined for prod build

### ISSUE THREAD

**Issue by Maks-Yaremenko** [sibiraj-s/ngx-editor#313]
# I'm submitting a
-  Bug report

## Description

_A bug:
On production, we are facing ERROR TypeError: Cannot read property 'execute' of undefined. 
On production whenever I'm clicking on the buttons on the editor I see no reaction but errors in the console.

![image](https://user-images.githubusercontent.com/22315741/104700275-937bd880-571c-11eb-9e2a-ad2a7c3761ef.png)

### Version Information

```
ngx-editor: 5.2.0
angular: 9.1.9
```

### Browser

-  Chrome - 87.0.4280.88 (Official Build) (x86_64)

**Comment by sibiraj-s** [comment]
Only In production?

**Comment by Maks-Yaremenko** [comment]
yes, I'm facing this issue only in production.

**Comment by sibiraj-s** [comment]
Can you provide a minor setup to reproduce this? I am not sure what config you have or what causes this. I tried enabling prodMode and everything works fine.

**Comment by Maks-Yaremenko** [comment]
I prepared minimal reproduction for you, https://github.com/Maks-Yaremenko/ngx-editor-prod-issue

you can only click on the buttons related to colors, link + image, other buttons are not working.

In case if you'll type smth in the editor you will see the next errors.
![image](https://user-images.githubusercontent.com/22315741/104717672-56badc00-5732-11eb-83bf-bc638d239143.png)

I'll really appreciate your help, cuz it's a blocker of our today release(((

**Comment by sibiraj-s** [comment]
Checking...

**Comment by Maks-Yaremenko** [comment]
Thank you, I'm doing the same as well.

**Comment by sibiraj-s** [comment]
<img width="824" alt="Screenshot 2021-01-15 at 5 29 32 PM" src="https://user-images.githubusercontent.com/20282546/104724849-449f6480-5757-11eb-84d1-f958af6735fd.png">

I have npm 6. I ran `npm install` then `npm start`. Getting this error.

**Comment by Maks-Yaremenko** [comment]
I'm using 7.0.10, that's weird

**Comment by Maks-Yaremenko** [comment]
I created this project using npx @angular/cli@9 new ngx-editor-prod-issue

**Comment by sibiraj-s** [comment]
Can you check and let me know I couldn't continue. I am not sure if npm version is an issue.

**Comment by Maks-Yaremenko** [comment]
could you provide the exact version of npm, I'll install it and will try to setup deps on my side, tested with the current version - no issues with deps

**Comment by Maks-Yaremenko** [comment]
![image](https://user-images.githubusercontent.com/22315741/104727584-3d1f9180-573e-11eb-9178-61caef133d08.png)
I had some errors during installation using npm -v === 3.10.10 this version was delivered with node v. 6.14.4, could you please specify the version on the node that you are using?

**Comment by Maks-Yaremenko** [comment]
installed node v14.12.0 which was delivered with npm v6.14.8. Installed all packages with no issues, but faced with the same issue during npm start that you are mentioned...
![image](https://user-images.githubusercontent.com/22315741/104728156-262d6f00-573f-11eb-9159-2a8c74d84b60.png)

**Comment by Maks-Yaremenko** [comment]
that's really weird.. but `npm install @angular/material@9.2.2` with npm start
![image](https://user-images.githubusercontent.com/22315741/104728661-e87d1600-573f-11eb-828e-abb459a5bed5.png)

**Comment by Maks-Yaremenko** [comment]
with the downgraded material library I'm able to start the app with the --prod flag, and reproduce the issue

**Comment by Maks-Yaremenko** [comment]
could you please reply whenever you'll be able to take a look?

**Comment by sibiraj-s** [comment]
> m -v === 3.10.10 this version was delivered with node v. 6.14.4, could you please specify the version on the node that you are using?

I am using node 14.15.1

**Comment by sibiraj-s** [comment]
> could you please reply whenever you'll be able to take a look?

Sorry. Today is a festive day. I am on a holiday. thats why the delay.

**Comment by Maks-Yaremenko** [comment]
my regular version is 15.2.0, with npm 7.0.10. But currently, I'm working with version node 14.12.0 and npm v6.14.8 as I stated above, to be closer to your environment.

**Comment by sibiraj-s** [comment]
> that's really weird.. but `npm install @angular/material@9.2.2` solves the issue with npm start

I did this and it didn't work.

**Comment by Maks-Yaremenko** [comment]
weird(((

**Comment by sibiraj-s** [comment]
I am trying now with npm 7

**Comment by sibiraj-s** [comment]
Works now. Not sure why npm is an issue here. Checking the issue.

**Comment by Maks-Yaremenko** [comment]
I even bumped Angular version to 10.2.4, and material to 10.2.7.. ng serve works fine, but ng serve --prod gives me the same errors as I mentioned in the description.

**Comment by sibiraj-s** [comment]
Checking on that.

**Comment by Maks-Yaremenko** [comment]
I figured out what caused after installation error... material requires cdk library, but I didn't install it. sorry for that
![image](https://user-images.githubusercontent.com/22315741/104734420-e4092b00-5748-11eb-9f63-576f6f21711f.png)

**Comment by sibiraj-s** [comment]
> that's really weird.. but `npm install @angular/material@9.2.2` solves the issue with npm start
> ![image](https://user-images.githubusercontent.com/22315741/104728661-e87d1600-573f-11eb-828e-abb459a5bed5.png)

NO issues. npm 7 installs peerDeps automatically. I saw the issue and try to figure out what causes this.

**Comment by Maks-Yaremenko** [comment]
yes, seems like. I just let you know what causes the issue with installation when u used v6 of npm.

**Comment by sibiraj-s** [comment]
I think I found the issue. Some issue with `Map`. 

Setting the value by

```js
const SimpleCommands = new Map();
SimpleCommands.set('bold', STRONG);
```
and getting it by  
```js
SimpleCommands.get('bold') // returns undefined.
```
returns undefined. Not sure if this is a problem with the library.

**Comment by Maks-Yaremenko** [comment]
probably I need to add Map polyfill in the app!

**Comment by sibiraj-s** [comment]
Polyfill is needed if the feature is not available. I works in development because it is supported. It shouldn't behave differently in development and production.

**Comment by sibiraj-s** [comment]
I will be unavailable for 30mins. Can you search around this. and also see if this is happening in angular 11 as well?

**Comment by Maks-Yaremenko** [comment]
I'm trying to migrate my code to your example with 11v of Angular. In progress. I'll let you know when I'll finish.

**Comment by sibiraj-s** [comment]
sure.

**Comment by sibiraj-s** [comment]
Also I found the `childCount` issue you mentioned while updating to v6. The regression was since 5.3

**Comment by Maks-Yaremenko** [comment]
yeah.. I mentioned that issue... in some of the other reports... which was one of the reasons why I didn't migrate to the latest version. FYO: seems like polyfills didn't help

**Comment by Maks-Yaremenko** [comment]
so, I moved my implementation to your example with v11... seems like no issues on angular 11... but issue reproducible on v9, v10

**Comment by sibiraj-s** [comment]
Strange.

**Comment by sibiraj-s** [comment]
Something is wrong with the compiler it seems. I am checking. Meanwhile can you update to v6.

> yeah.. I mentioned that issue... in some of the other reports... It was one of the reasons why I didn't migrate to the latest version. 

It is fixed in v6.beta.2 Update should be straight forward. It will be easy for me to work and release with v6.

**Comment by sibiraj-s** [comment]
@Maks-Yaremenko

Setting.

```
aot: false
# and 
buildOptimizer: false 
```

in angular.json fixes the issue

**Comment by Maks-Yaremenko** [comment]
currently, I'm working in another direction - bumping the Angular version... to 11. Hope nothing will be broken. Thank you for the suggestion

**Comment by sibiraj-s** [comment]
Consider the fallback if things go south.

**Comment by Maks-Yaremenko** [comment]
Thank you!

**Comment by sibiraj-s** [comment]
Update. 
Just setting 
```
buildOptimizer: false 
```
fixes the issue.

**Comment by sibiraj-s** [comment]
Were you able to resolve the issue?

**Comment by Maks-Yaremenko** [comment]
Hi! Sorry for the late response, disabling build optimization wasn't an option for us. We decide to update our project to use the latest Angular version, where your package works fine. So after this update, it works perfectly for us. Anyhow issue on Angular v9 still reproducable.

**Comment by sibiraj-s** [comment]
Cool. Glad I could assit.

I will try to fix the issue though meanwhile. I guess the transition to ivy based compiler in angular 9 has some issues.

**Comment by sibiraj-s** [comment]
Later I checked it not the Map. I tried it with simple objects as well. but didn't work. Some how 

```ts
const myClass = new Class()
```
myClass is becoming undefined. Not sure why.

**Comment by Maks-Yaremenko** [comment]
I see! I really love to investigate such issues, looks like a challenge) I believe it would very interesting for you to find out what is going on)

**Comment by sibiraj-s** [comment]
Yeah. Sadly I just couldn't find enough time to spend on this. I will try this weekend. Meanwhile I would gladly accept any kind of help if possible.

**Comment by Maks-Yaremenko** [comment]
Apologize, but currently, I'm reassigned to another project, so have to spend all my capacity to get into it.

**Comment by sibiraj-s** [comment]
No Issues.

**Comment by sibiraj-s** [comment]
I was able to narrow it down to the 

```
sideEffects: false
```

added by `ng-packager` to package.json during build time. Setting it to true or removing it fixes that.

**Comment by Maks-Yaremenko** [comment]
thank you for the info!

**Comment by sibiraj-s** [comment]
Fixed in v7.0.3.
